### DISCUSSION TITLE
Make impersonate flexibility as powerful as KABBOUCHI/nova-impersonate

### DISCUSSION THREAD

**Discussion by ziming** [laravel/nova-issues#4141] [Q&A] [answered]
Currently the impersonate feature in Nova is not as powerful as the package it took inspiration from

https://github.com/KABBOUCHI/nova-impersonate

In the above package, the method signature is like this

```php
use \Illuminate\Contracts\Auth\Authenticatable;
public function canImpersonate($impersonated = null): bool // Not sure if I can type hint $impersonated with Authenticatable
    {
       
    }

    public function canBeImpersonated(Authenticatable $impersonator = null): bool
    {

    }
```

In Nova Impersonate trait, the 2 methods have the same name but they currently accept no arguments. Thus less flexibility to decide when to allow impersonation

```php
public function canImpersonate()
    {
        return Gate::forUser($this)->check('viewNova');
    }


    public function canBeImpersonated()
    {
        return true;
    }
```

An example use case of why I want it to be the same as KABBOUCHI/nova-impersonate

```php
public function canImpersonate($impersonated = null): bool // Not sure if I can type hint $impersonated with Authenticatable
    {
       return $this->rank > $impersonated->rank;
    }
```

**Comment by crynobone** [comment, marked as answer]
You may use https://github.com/KABBOUCHI/nova-impersonate

**Reply by ziming** [reply]
I see Nova 4 support is added just today in that package.

Would it be too hard for Official Nova to just pass in the optional argument (impersonated/impersonator) so we can avoid installing 1 more 3rd party package?
